src/guiguts/spell.py

```diff
@@ -0,0 +1,326 @@
+        # If it's a bad word or in the dictionary, we're done
+        if word_status == SPELL_CHECK_OK_BAD or word_status == SPELL_CHECK_OK_YES:
+            return word_status
+
+        # Otherwise, try swapping apostrophes if there are any, and re-check
+        if "'" in word:
+            word = re.sub(r"'", r"’", word)
+        elif "’" in word:
+            word = re.sub(r"’", r"'", word)
+        else:
+            return word_status
+        return self.spell_check_word_case(word)
+
+    def spell_check_word_case(self, word: str) -> int:
+        """Spell check word, allowing for differences of case.
+
+        Check same case, lower case, or title case (e.g. LONDON matches London)
+        Can't just do case-insensitive check because we don't want "london" to be OK.
+
+        Args:
+            word: Word to be checked.
+
+        Returns:
+            SPELL_CHECK_OK_BAD if word is in the bad_words list
+            SPELL_CHECK_OK_YES if word is in dictionary
+            SPELL_CHECK_OK_NO if neither (wrong spelling)
+        """
+        lower_word = word.lower()
+        if lower_word == word:
+            lower_word == ""
+        # Caution converting to title case - we want "LONDON" to match "London",
+        # but we don't want "london" to match London, so leave first letter case
+        # as-is, and lowercase the rest
+        if len(word) > 1:
+            title_word = word[0:1] + word[1:].lower()
+            if title_word == word:
+                title_word == ""
+        else:
+            title_word = ""
+
+        # Todo: check if word, lower_word or title_word are bad words
+
+        # Now check dictionary for good words
+        if (
+            word in self.dictionary
+            or (lower_word and lower_word in self.dictionary)
+            or (title_word and title_word in self.dictionary)
+        ):
+            return SPELL_CHECK_OK_YES
+
+        return SPELL_CHECK_OK_NO
+
+    def add_words_from_dictionary(self, filename: str) -> bool:
+        """Loads given dictionary into SpellChecker dict.
+
+        Args:
+            filename: Full pathname of dictionary to load.
+
+        Returns:
+            True if dictionary file exists.
+        """
+        try:
+            with open(filename, "r", encoding="utf-8") as f:
+                for line in f:
+                    word = line.strip()
+                    self.dictionary[word] = True
+            return True
+        except FileNotFoundError:
+            return False
+
+    def add_words_from_language(self, lang: str) -> None:
+        """Add dictionary words for given language.
+
+        Attempts to load default and user dictionaries.
+        Outputs warning if default dictionary doesn't exits.
+        Raises DictionaryNotFoundError exception if neither dictionary exists.
+
+        Args:
+            lang: Language to be loaded.
+        """
+        filename = str(DEFAULT_DICTIONARY_DIR.joinpath(f"dict_{lang}_default.txt"))
```

Comment:
The casting as a string and then using a standard open forces this back into a filesystem, not package, access method. The `DEFAULT_DIRECTORY_DIR` should be a Traversable object that we can `.open()` directly. This way if the package is being read and used from within a `.tar.gz` this still works.

Reply:
Ahhhh. That makes sense. I did that because the routine I used to load the default dictionary (bundled with the system) and the user's dictionary (in their GGprefs dir) took a filename string. I'll adapt it - I can probably get a Traversable for the user's dictionary too if I have a look round.

Reply:
I found that `Traversable` is like the `Path` class, so am using the latter for the user's dictionary. Then the method that opens and loads the dictionary can take `Traversable|Path` as its input argument type, as they both support `.open()`